File: data_structures/graph/python/graph.py
from __future__ import annotations
from collections import deque


class Node:
    """Graph node used by problem solutions (adjacency via pointers)."""
    def __init__(self, val: int = 0, neighbors: list[Node] | None = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


class Graph:
    """Graph using adjacency list representation."""
    def __init__(self, directed: bool = False):
        self._adjacency: dict[int, list[int]] = {}
        self._directed = directed

    def add_vertex(self, v: int) -> None:
        if v not in self._adjacency:
            self._adjacency[v] = []

    def add_edge(self, from_v: int, to_v: int) -> None:
        self.add_vertex(from_v)
        self.add_vertex(to_v)
        self._adjacency[from_v].append(to_v)
        if not self._directed:
            self._adjacency[to_v].append(from_v)

    def remove_edge(self, from_v: int, to_v: int) -> None:
        if from_v in self._adjacency:
            try:
                self._adjacency[from_v].remove(to_v)
            except ValueError:
                pass
        if not self._directed and to_v in self._adjacency:
            try:
                self._adjacency[to_v].remove(from_v)
            except ValueError:
                pass

    def has_edge(self, from_v: int, to_v: int) -> bool:
        return from_v in self._adjacency and to_v in self._adjacency[from_v]

    def neighbors(self, v: int) -> list[int]:
        return self._adjacency.get(v, [])

    def vertices(self) -> list[int]:
        return sorted(self._adjacency.keys())

    def bfs(self, start: int) -> list[int]:
        if start not in self._adjacency:
            return []
        visited: set[int] = {start}
        order: list[int] = []
        queue: deque[int] = deque([start])
        while queue:
            current = queue.popleft()
            order.append(current)
            for neighbor in sorted(self._adjacency[current]):
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
        return order

    def dfs(self, start: int) -> list[int]:
        if start not in self._adjacency:
            return []
        visited: set[int] = set()
        order: list[int] = []
        self._dfs_helper(start, visited, order)
        return order

    def _dfs_helper(self, v: int, visited: set[int], order: list[int]) -> None:
        visited.add(v)
        order.append(v)
        for neighbor in sorted(self._adjacency[v]):
            if neighbor not in visited:
                self._dfs_helper(neighbor, visited, order)

    def shortest_path(self, start: int, end: int) -> tuple[list[int] | None, int]:
        if start not in self._adjacency or end not in self._adjacency:
            return None, -1
        if start == end:
            return [start], 0
        visited: set[int] = {start}
        parent: dict[int, int] = {}
        queue: deque[int] = deque([start])
        while queue:
            current = queue.popleft()
            for neighbor in self._adjacency[current]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    parent[neighbor] = current
                    if neighbor == end:
                        path = self._reconstruct_path(parent, start, end)
                        return path, len(path) - 1
                    queue.append(neighbor)
        return None, -1

    def _reconstruct_path(self, parent: dict[int, int], start: int, end: int) -> list[int]:
        path = []
        current = end
        while current != start:
            path.append(current)
            current = parent[current]
        path.append(start)
        path.reverse()
        return path

File: data_structures/graph/graph.go
package datastructures

import "sort"

// Graph using adjacency list representation
type Graph struct {
	adjacency map[int][]int
	directed  bool
}

// NewGraph creates a new graph. If directed is true, edges are one-directional.
func NewGraph(directed bool) *Graph {
	return &Graph{
		adjacency: make(map[int][]int),
		directed:  directed,
	}
}

// AddVertex adds a vertex with an empty adjacency list if it does not already exist.
func (g *Graph) AddVertex(v int) {
	if _, exists := g.adjacency[v]; !exists {
		g.adjacency[v] = []int{}
	}
}

// AddEdge adds an edge from 'from' to 'to'. For undirected graphs, the reverse
// edge is also added. Vertices are auto-created if they do not exist.
func (g *Graph) AddEdge(from, to int) {
	g.AddVertex(from)
	g.AddVertex(to)
	g.adjacency[from] = append(g.adjacency[from], to)
	if !g.directed {
		g.adjacency[to] = append(g.adjacency[to], from)
	}
}

// RemoveEdge removes the edge from 'from' to 'to'. For undirected graphs, the
// reverse edge is also removed.
func (g *Graph) RemoveEdge(from, to int) {
	g.adjacency[from] = removeFromSlice(g.adjacency[from], to)
	if !g.directed {
		g.adjacency[to] = removeFromSlice(g.adjacency[to], from)
	}
}

// removeFromSlice removes the first occurrence of val from the slice.
func removeFromSlice(slice []int, val int) []int {
	for i, v := range slice {
		if v == val {
			return append(slice[:i], slice[i+1:]...)
		}
	}
	return slice
}

// HasEdge returns true if an edge exists from 'from' to 'to'.
func (g *Graph) HasEdge(from, to int) bool {
	neighbors, exists := g.adjacency[from]
	if !exists {
		return false
	}
	for _, v := range neighbors {
		if v == to {
			return true
		}
	}
	return false
}

// Neighbors returns the list of adjacent vertices for a given vertex.
func (g *Graph) Neighbors(v int) []int {
	return g.adjacency[v]
}

// Vertices returns a sorted list of all vertices in the graph.
func (g *Graph) Vertices() []int {
	vertices := make([]int, 0, len(g.adjacency))
	for v := range g.adjacency {
		vertices = append(vertices, v)
	}
	sort.Ints(vertices)
	return vertices
}

// BFS performs a breadth-first search starting from the given vertex and returns
// the visit order.
func (g *Graph) BFS(start int) []int {
	if _, exists := g.adjacency[start]; !exists {
		return nil
	}

	visited := make(map[int]bool)
	order := []int{}
	queue := []int{start}
	visited[start] = true

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		order = append(order, current)

		neighbors := make([]int, len(g.adjacency[current]))
		copy(neighbors, g.adjacency[current])
		sort.Ints(neighbors)

		for _, neighbor := range neighbors {
			if !visited[neighbor] {
				visited[neighbor] = true
				queue = append(queue, neighbor)
			}
		}
	}

	return order
}

// DFS performs a depth-first search starting from the given vertex and returns
// the visit order.
func (g *Graph) DFS(start int) []int {
	if _, exists := g.adjacency[start]; !exists {
		return nil
	}

	visited := make(map[int]bool)
	order := []int{}
	g.dfsHelper(start, visited, &order)
	return order
}

func (g *Graph) dfsHelper(v int, visited map[int]bool, order *[]int) {
	visited[v] = true
	*order = append(*order, v)

	neighbors := make([]int, len(g.adjacency[v]))
	copy(neighbors, g.adjacency[v])
	sort.Ints(neighbors)

	for _, neighbor := range neighbors {
		if !visited[neighbor] {
			g.dfsHelper(neighbor, visited, order)
		}
	}
}

// ShortestPath finds the shortest path between start and end using BFS (for
// unweighted graphs). It returns the path as a slice of vertices and the
// distance. If no path exists, it returns (nil, -1).
func (g *Graph) ShortestPath(start, end int) ([]int, int) {
	if _, exists := g.adjacency[start]; !exists {
		return nil, -1
	}
	if _, exists := g.adjacency[end]; !exists {
		return nil, -1
	}
	if start == end {
		return []int{start}, 0
	}

	visited := make(map[int]bool)
	parent := make(map[int]int)
	queue := []int{start}
	visited[start] = true

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		for _, neighbor := range g.adjacency[current] {
			if !visited[neighbor] {
				visited[neighbor] = true
				parent[neighbor] = current
				if neighbor == end {
					return reconstructPath(parent, start, end)
				}
				queue = append(queue, neighbor)
			}
		}
	}

	return nil, -1
}

// reconstructPath builds the path from start to end using the parent map.
func reconstructPath(parent map[int]int, start, end int) ([]int, int) {
	path := []int{}
	current := end
	for current != start {
		path = append(path, current)
		current = parent[current]
	}
	path = append(path, start)

	// Reverse the path
	for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
		path[i], path[j] = path[j], path[i]
	}

	return path, len(path) - 1
}
